ion-c-sys/src/lib.rs

```diff
@@ -306,6 +311,315 @@ mod test {
     }
 }
 
+/// Creates an operating context for decNum with maximum ranges primarily for conversions
+/// between `ION_DECIMAL` and `BigDecimal`.
+///
+/// This range is still below the capabilities of `BigDecimal`
+#[inline]
+fn make_context() -> decContext {
+    let mut ctx = decContext::default();
+    unsafe { decContextDefault(&mut ctx, DEC_INIT_DECQUAD as i32) };
+    ctx.digits = DEC_MAX_DIGITS as i32;
+    ctx.emax = DEC_MAX_EMAX as i32;
+    ctx.emin = DEC_MIN_EMIN as i32;
+    // make sure we don't pad decQuads
+    ctx.clamp = 0;
+
+    ctx
+}
+
+const DEC_QUAD_DIGITS: u64 = 37;
+
+impl ION_DECIMAL {
+    /// Assigns a `BigDecimal` into this `ION_DECIMAL`.
+    pub fn try_assign_bigdecimal(&mut self, value: &BigDecimal) -> IonCResult<()> {
+        let digits = value.digits();
+
+        // FIXME amzn/ion-rust#80 - this breaks encapsulation
+        ionc!(ion_decimal_free(self))?;
+        if digits > DEC_QUAD_DIGITS {
+            self.type_ = ION_DECIMAL_TYPE_ION_DECIMAL_TYPE_NUMBER;
+            // need to allocate the number field
+            ionc!(_ion_decimal_number_alloc(
+                ptr::null_mut(),
+                digits.try_into()?,
+                &mut self.value.num_value,
+            ))?;
+        } else {
+            self.type_ = ION_DECIMAL_TYPE_ION_DECIMAL_TYPE_QUAD;
+        }
+        ionc!(ion_decimal_zero(self))?;
+
+        let mut ctx = make_context();
+        let (coefficient, scale) = value.as_bigint_and_exponent();
+
+        // set the coefficient
+        let mut ion_coefficient = IonIntPtr::try_from_bigint(&coefficient)?;
+        ionc!(ion_decimal_from_ion_int(
+            self,
+            &mut ctx,
+            ion_coefficient.as_mut_ptr()
+        ))?;
+
+        // scale the value
+        let mut dec_scale = IonDecimalPtr::try_from_i32((-scale).try_into()?)?;
+        ionc!(ion_decimal_scaleb(
+            self,
+            self,
+            dec_scale.as_mut_ptr(),
+            &mut ctx
+        ))?;
+
+        match unsafe { decContextGetStatus(&mut ctx) } {
+            0 => Ok(()),
+            // FIXME amzn/ion-rust#78 - we need to be more careful around the limits of decQuad.
+            DEC_Inexact => Err(IonCError::from(ion_error_code_IERR_NUMERIC_OVERFLOW)),
+            // FIXME amzn/ion-rust#79 - we need to fix decNumber support.
+            DEC_Invalid_context => Err(IonCError::from(ion_error_code_IERR_INVALID_STATE)),
+            _ => Err(IonCError::from(ion_error_code_IERR_INTERNAL_ERROR)),
+        }
+    }
+
+    /// Converts this `ION_DECIMAL` to a `BigDecimal`.
+    ///
+    /// Special decimal values such as NaN and infinity are not supported for conversion.
+    ///
+    /// This implementation borrows mutably, to avoid a copy of the underlying
+    /// decimal implementation, but does not change the value.
```

Comment:
Can you expand on this? Why is the `&mut` necessary if the value isn't being changed? Tough to tell at a glance without an IDE, but maybe one of the nested function calls demands it?

Reply:
The value is being changed because I am scaling it from/to the exponent to extract the coefficient.  I can add a comment to this effect, but I am doing the operation in place.

Reply:
Added more documentation about this.